### PR TITLE
[Codegen][cuda-fp16] fallback to fp32 simulation when cuda arch < sm53

### PR DESCRIPTION
**Pull request by yzhliu**
This ensures fp16 can still be computed on old devices.
The code is borrowed from mshadow.

An immediate usage will be tvm op support in mxnet - mxnet is able to run fp16 on device which does not support fp16 natively.

@vinx13 @tqchen @Laurawly @reminisce Please advise.

### CODE REVIEW

**Comment by tqchen** [comment]
I am not sure if we really want to enable fp16 on devices that does not have native support. never-the-less. I think we can enable it. However, it would be great to separate this dep lib into a separate const string var that uses string literal to include the files. So we won't have a long string in the content.

**Comment by vinx13** [comment]
I agree better we can put the code in a separate file instead of string concatenation. The code itself is good

**Comment by yzhliu** [comment]
@vinx13 @tqchen please review again

File: src/codegen/codegen_cuda.cc [hunk 1]
```diff
@@ -50,6 +50,9 @@ void CodeGenCUDA::AddFunction(LoweredFunc f) {
 
 std::string CodeGenCUDA::Finish() {
   if (enable_fp16_) {
+    static constexpr const char* _cuda_half_t_def =
+    #include "literal/cuda_half_t.txt"
```
Comment (tqchen) [inline src/codegen/codegen_cuda.cc]: consider still keep it as a header file so we can apply the license header, define a global constant

Comment (yzhliu) [reply src/codegen/codegen_cuda.cc]: @tqchen addressed.

**Comment by tqchen** [comment]
Thanks @yzhliu !

**Pull request merged**